### ISSUE TITLE
Add lint for ensuring a JSDoc exists for public (exported) APIs

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno_lint#970]
Ref https://github.com/denoland/deno_std/issues/650

It's useful to automatically ensure that all public APIs have appropriate JSDocs, I believe an additional lint rule is the way to go for this purpose.

**Comment by Skillz4Killz** [comment]
As a library developer this is something really important to me. I stopped using deno lint a while ago because of the headaches i had just deno-lint-ignoring so many lines. However, having proper documentation would be worth that trouble because it helps save me time since anything undocumented means i have to explain things to users when they ask how something works.

**Ideal Functionality:**

```ts
// PASS
/** does something */

// PASS BUT NOT REQUIRED
/**
  * does something
  * @param bar string a string to run this function on
  * @returns string
*/

// FAIL
/** */

// FAIL
// does something

// FAIL SINCE NO COMMENT 
export function foo(bar: string) {
  return bar;
}
```

I would prefer`/** does something */` be considered enough because i use TS and i believe the types + clear param names are enough and extra docs on them is just redundant and lowers productivity. However, i would suggest an option for this since JS devs might want this.

**Comment by kitsonk** [comment]
Do we expect this to be on by default?

This can quickly get very annoying (like when we had the lint rule about requiring a return type annotation).  For example if I write a utility function in a utility module, it isn't part of my public API, but I export it, and if it is well named and has well named parameters, JSDoc'ing that function or method is _useless_.

**Comment by bartlomieju** [comment]
> Do we expect this to be on by default?
> 
> This can quickly get very annoying (like when we had the lint rule about requiring a return type annotation). For example if I write a utility function in a utility module, it isn't part of my public API, but I export it, and if it is well named and has well named parameters, JSDoc'ing that function or method is _useless_.

In the beginning definitely yes, but I would love to have it on for `deno` repo by default. Ideally we could land on a solution that doesn't involve config, but I'm afraid in this case configuration might be inevitable.

**Comment by kitsonk** [comment]
I am :-1: for having it if it is on be default.  I can see it useful for `std`, but for end users, it would likely make things too annoying.

**Comment by bartlomieju** [comment]
> I am 👎 for having it if it is on be default. I can see it useful for `std`, but for end users, it would likely make things too annoying.

Okay sure, let's add the rule and try it out on some real life projects before deciding one way or the other.

**Comment by iuioiua** [comment]
This sounds like a great way of enforcing documentation. Along the same lines, I think a cool to enforce omitting types from JSDoc tags like `@arg` and `@returns` within TypeScript files, as it's not needed. Ref:
* denoland/deno_std#2493
* denoland/deno_std#2515
